PermissionVault/contracts/PermissionVault.sol

```diff
@@ -0,0 +1,348 @@
+  /*                                  Modifiers                                 */
+  /* -------------------------------------------------------------------------- */
+
+  modifier onlyController() {
+    if (!hasRole(CONTROLLER_ROLE, msg.sender)) {
+      revert NotController();
+    }
+    _;
+  }
+
+  /* -------------------------------------------------------------------------- */
+  /*                             External Functions                             */
+  /* -------------------------------------------------------------------------- */
+
+  /**
+   * @dev Constructor
+   */
+  constructor() {
+    _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    _setupRole(CONTROLLER_ROLE, msg.sender);
+  }
+
+  /**
+   * @notice The owner can use this function to grant the CONTROLLER_ROLE to a
+   * new address. The controller parameter is the address that will be granted
+   * the CONTROLLER_ROLE.
+   * @dev Callable by owner
+   * @param controller new controller address
+   */
+  function addController(address controller) external onlyOwner {
+    _setupRole(CONTROLLER_ROLE, controller);
+
+    emit AddController(controller);
+  }
+
+  /**
+   * @notice The owner can use this function to revoke the CONTROLLER_ROLE from
+   * an address. The controller parameter is the address that will have the
+   * CONTROLLER_ROLE revoked.
+   * @dev Callable by owner
+   * @param controller controller address
+   */
+  function removeController(address controller) external onlyOwner {
+    revokeRole(CONTROLLER_ROLE, controller);
+
+    emit RemoveController(controller);
+  }
+
+  /**
+   * @notice A controller can use this function to deposit Ether into the
+   * contract.
+   */
+  function depositEther() external payable {
+    if (msg.value == 0) {
+      revert ZeroAmount();
+    }
+
+    emit DepositEther(msg.sender, msg.value);
+  }
+
+  /**
+   * @notice A controller can use this function to withdraw Ether from the contract.
+   * @dev Callable by controller
+   * @param to the address that will receive the withdrawn Ether
+   * @param amount the amount of Ether to withdraw
+   */
+  function withdrawEther(
+    address to,
+    uint256 amount
+  ) external nonReentrant onlyController {
+    if (amount > address(this).balance) {
+      revert NotEnoughBalance();
+    }
+
+    (bool success, ) = to.call{value: amount}("");
+    require(success, "Transfer ether failed");
+
+    emit WithdrawEther(msg.sender, to, amount);
+  }
+
+  /**
+   * @notice A controller can use this function to deposit ERC20 tokens into the contract.
+   * @param token the ERC20 token contract address
+   * @param amount the amount of tokens to deposit
+   */
+  function depositERC20(IERC20 token, uint256 amount) external {
+    if (amount == 0) {
+      revert ZeroAmount();
+    }
+    if (token.balanceOf(msg.sender) < amount) {
+      revert NotEnoughBalance();
+    }
+
+    token.safeTransferFrom(msg.sender, address(this), amount);
```

Comment:
Please check allowance amount before transferFrom

Reply:
don't need it

Reply:
don't need it

Reply:
agreee, oz already had it